Install external repos

ibrahimhroob opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

-->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
Add option to install external repos inside src ws

## Related Tickets & Documents

<!--
For pull requests that relate or close an issue, please include them
below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "closes #1234" would connect the current pull
request to issue 1234.  And when we merge the pull request, Github will
automatically close the issue.
-->

- Related Issue #
- Closes #

## QA Instructions, Screenshots, Recordings

_Please replace this line with instructions on how to test your changes._

## [optional] Are there any post deployment tasks we need to perform?

ibrahimhroob committed:
update docker file

ibrahimhroob committed:
add bash script to install external repos

ibrahimhroob commented:
@Cyano0 can you please merge this once finished the checks, this now will be able to install the external repos inside the src

ibrahimhroob commented:
@marc-hanheide is this the correct way to add external repos to the package?

marc-hanheide commented on .devcontainer/user_install.dockerfile:
```diff
@@ -0,0 +1,22 @@
+COPY .devcontainer/scripts/install_exterenal_repos.sh ${WORKSPACES}/src/install_exterenal_repos.sh
+
+# Make the script executable and run it
+RUN chmod +x ${WORKSPACES}/src/install_exterenal_repos.sh && \
```
typo ;-)

marc-hanheide commented on .devcontainer/Dockerfile:
```diff
@@ -1,31 +1,52 @@
 FROM depbuilder as final
+
+# Copy the full src directory into the final stage
+COPY ./src ${WORKSPACES}/src
```
I'm not sure I fully understand what the aim is here. If it's for a devcontainer for this repo, then copying in the source here seems not what is intended. This appears to be a bit of a mix between creating a fully deployable container, and installing dependencies for a dev container.

ibrahimhroob replied:
Because we are having multi stage build, when installing the dependencies at the first stage, they do not show in the final stage, so I have to copy it. I am sure this is not the correct way to do it!!

marc-hanheide replied:
what do you mean, they do not show in the final stage? The decencies are resolved and part of the image. Here you are copy the source from the repositiory not the dependencies. For a devcontainer you don't copy the source as the repositiory itself is mounted in the devcontainer. So copying them only makes sense when and if one wants to make an independent docker file, to be used NOT as a devcontainer.

marc-hanheide replied:
what do you mean, they do not show in the final stage? The decencies are resolved and part of the image. Here you are copy the source from the repositiory not the dependencies. For a devcontainer you don't copy the source as the repositiory itself is mounted in the devcontainer. So copying them only makes sense when and if one wants to make an independent docker file, to be used NOT as a devcontainer.

ibrahimhroob replied:
sorry I miss presented the issue. The problem that I am facing is when installing external repos in the `src` folder using this script https://github.com/ibrahimhroob/aoc_hri_ros2_pkg/blob/a2b746b5b3a1ca2ede081df32c41924266504edb/.devcontainer/user_install.dockerfile#L18-L20 they do not show after launching the devcontainer, when building it shows that it has cloned the external repos into `src` folder, however, after building the container we do not see the cloned repo in the `src` directory.

ibrahimhroob replied:
How can I add external packages inside the src folder? The template repository is designed to handle dependency installation through a multi-stage Docker build process, which works efficiently for the existing setup. However, what is the best practice for including additional packages within the src folder that the package is dependent on and they do not have ros dep package?

marc-hanheide commented on .devcontainer/Dockerfile:
```diff
@@ -1,31 +1,52 @@
 ENV DEBIAN_FRONTEND=noninteractive
 
+# Set the LD_PRELOAD environment variable
+ENV LD_PRELOAD=/usr/lib/libdlfaker.so:/usr/lib/libvglfaker.so
```
Do we know if this is required. I don't think so

ibrahimhroob replied:
It is not required, but it is to remove the annoying warning

marc-hanheide commented on .devcontainer/Dockerfile:
```diff
@@ -1,31 +1,52 @@
+ENV LD_PRELOAD=/usr/lib/libdlfaker.so:/usr/lib/libvglfaker.so
+
+# Define the workspace path
+ENV WORKSPACES=/workspaces/aoc_hri_ros2_pkg
```
Same here, for a devcontainer, I don't know why needed.

ibrahimhroob replied:
to remove this warning:
ros@f1a7eecc10f4:~/aoc_hri_ws$ sudo ls 
ERROR: ld.so: object 'libdlfaker.so' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
ERROR: ld.so: object 'libvglfaker.so' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.

ibrahimhroob committed:
make sure installing external repos works

Cyano0 committed:
Changes to Docker file

ibrahimhroob committed:
update post-create-command and added error detection in post-create.sh

ibrahimhroob committed:
adding force and recursive flags

ibrahimhroob committed:
Install Python dependencies for the ultralytics_ros package

ibrahimhroob committed:
ultralytics_ros using the normal way due to this:   Could not init/update submodules: fatal: No url found for submodule path 'src/ultralytics_ros' in .gitmodules

ibrahimhroob committed:
update ros-ci to Fetches submodules during checkout

ibrahimhroob committed:
Fix CI pipeline submodule issue by installing Git 2.18+ to enable full repository clone

ibrahimhroob committed:
Fix CI: Remove sudo to run Git installation in container environment

ibrahimhroob committed:
Prevents attempting to fetch submodules

ibrahimhroob committed:
remove submodule

ibrahimhroob committed:
ci: set FORCE_JAVASCRIPT_ACTIONS_TO_NODE20 environment variable for Node.js compatibility

- Added FORCE_JAVASCRIPT_ACTIONS_TO_NODE20=true in CI workflow to ensure compatibility with Node.js 20
- This change supports updated GitHub Actions requirements and ensures consistency across the CI environment

ibrahimhroob committed:
Fix CI pipeline: Remove recursive submodule import for non-existent submodules

ibrahimhroob committed:
wip

ibrahimhroob committed:
remove submodule ref

ibrahimhroob committed:
ros-ci: revert to init state

ibrahimhroob committed:
update readme

ibrahimhroob committed:
Merge pull request #1 from Cyano0/dev

update with the latest changes

ibrahimhroob committed:
update on the merged comment to make it work

ibrahimhroob committed:
WIP

ibrahimhroob committed:
remove commented code

ibrahimhroob committed:
added bashrc commands

ibrahimhroob committed:
update CI devcontainer

ibrahimhroob committed:
copy external repos

ibrahimhroob committed:
Remove pip cache once all packages are installed to free up space

ibrahimhroob committed:
Remove unnecessary files or temporary files created during the setup

ibrahimhroob committed:
run on lcas

ibrahimhroob committed:
install expernal repo at root level

ibrahimhroob committed:
CI: devcontainer set to run on ubuntu-latest

ibrahimhroob committed:
bug fix

ibrahimhroob committed:
fix permissoins

ibrahimhroob committed:
set to rw

ibrahimhroob committed:
CI: update path to config file

ibrahimhroob committed:
blind guiding blind

ibrahimhroob committed:
CI updates

ibrahimhroob committed:
CI testing

ibrahimhroob committed:
CI: runs on LCAS

ibrahimhroob commented:
@Cyano0 please merge if you are happy with the changes
